### PR TITLE
wire: wire.Build now returns an error if it has any unused arguments

### PR DESCRIPTION
**Pull request by vangent**

### CODE REVIEW

File: wire/internal/wire/testdata/UnusedInputs/foo/foo.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Google LLC
```
Comment (zombiezen) [inline wire/internal/wire/testdata/UnusedInputs/foo/foo.go]: (For the directory, not this particular file/line:) Wire uses the term "input" to mean arguments passed to the injector. A better name for this test would be `UnusedProviders`.

File: wire/internal/wire/testdata/UnusedInputs/foo/wire.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/google/go-cloud/wire"
+)
+
+func injectBar() Bar {
+	wire.Build(
+		provideFoo,                            // needed as input for provideBar
+		provideBar,                            // needed for Bar
+		provideUnused,                         // not needed -> error
+		wire.Value("unused"),                  // not needed -> error
+		wire.NewSet(provideUnusedInSet),       // nothing in set is needed -> error
```
Comment (zombiezen) [inline wire/internal/wire/testdata/UnusedInputs/foo/wire.go]: Nice, this covers this quite thoroughly! Two more cases to check:

1. A set that is partially used (i.e. contains at least one provider that is used and at least one provider that is not) should not produce an error.
2. A set that is fully used should not produce an error.

File: wire/internal/wire/analyze.go [hunk 3]
```diff
@@ -226,14 +230,71 @@ dfs:
...
+	if len(ec.errors) > 0 {
+		return nil, ec.errors
+	}
 	return calls, nil
 }
 
-// buildProviderMap creates the providerMap field for a given provider set.
-// The given provider set's providerMap field is ignored.
-func buildProviderMap(fset *token.FileSet, hasher typeutil.Hasher, set *ProviderSet) (*typeutil.Map, []error) {
+func verifyArgsUsed(set *ProviderSet, used []*providerSetSrc, ec *errorCollector) {
```
Comment (zombiezen) [inline wire/internal/wire/analyze.go]: Instead of taking in an `*errorCollector`, return `[]error`.

**Comment by zombiezen** [comment]
(Fixes #164)

File: wire/internal/wire/analyze.go (hunk 3 above)
Comment (zombiezen) [inline wire/internal/wire/analyze.go]: Add a doc comment for this function.

File: wire/internal/wire/testdata/UnusedInputs/foo/foo.go (hunk 1 above)
Comment (vangent) [reply wire/internal/wire/testdata/UnusedInputs/foo/foo.go]: Done.

File: wire/internal/wire/analyze.go (hunk 3 above)
Comment (vangent) [reply wire/internal/wire/analyze.go]: Done.

Comment (vangent) [reply wire/internal/wire/analyze.go]: Done.

**Comment by vangent** [comment]
> The srcMap is being built for every provider set, where it only needs to be built for the top-level.

Hmm, any tips on how to detect this? By "top level", do you mean "only provider sets that are created for wire.Build, not for wire.NewSet"? 

Maybe add a boolean "retainSources" or some such to processNewSet?

> ... the salient information is giving the output type of bindings and values, since that will help the user more quickly identify what needs to be removed.

Done.

File: wire/internal/wire/analyze.go [hunk 4]
```diff
@@ -230,14 +230,15 @@ dfs:
...
+	if errs := verifyArgsUsed(set, used); len(errs) > 0 {
+		return nil, errs
 	}
 	return calls, nil
 }
 
-func verifyArgsUsed(set *ProviderSet, used []*providerSetSrc, ec *errorCollector) {
+// verifyArgsUsed ensures that all of the arguments in set were used during solve.
+func verifyArgsUsed(set *ProviderSet, used []*providerSetSrc) []error {
+	var errs []error
```
[the remark this answers is not in the document]
Comment (vangent) [reply wire/internal/wire/analyze.go]: Ack. Doesn't seem necessary to me.

File: wire/internal/wire/testdata/UnusedProviders/out.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+ERROR
+unused provider set
```
[the remark this answers is not in the document]
Comment (vangent) [reply wire/internal/wire/testdata/UnusedProviders/out.txt]: Do provider sets have names? imp.Name seems to be the empty string in my testcase. wire.NewSet(...) doesn't have any way to set a name, does it?

[the remark this answers is not in the document]
Comment (zombiezen) [reply wire/internal/wire/testdata/UnusedProviders/out.txt]: Provider sets defined as package variables should have this property, but it's possible this is broken.

[the remark this answers is not in the document]
Comment (vangent) [reply wire/internal/wire/testdata/UnusedProviders/out.txt]: OK. I will include the name if it's present and look into why it's the empty string in my test separately.

**Comment by vangent** [comment]
Per offline discussion, I think it's better to just always generate srcMap; it may be useful for other things like https://github.com/google/go-cloud/issues/165, and adding making it conditional adds a bunch of conditionals and "this field is sometimes set and sometimes nil" complexity.

File: wire/internal/wire/testdata/UnusedProviders/out.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (vangent) [reply wire/internal/wire/testdata/UnusedProviders/out.txt]: Done.

**Comment by zombiezen** [comment]
Cool. Thanks for opening #277.

**Pull request merged**